### PR TITLE
Remove conn.mux

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
conn.connReader() dispatches frames directly to sessions now.
Added conn.NextSession() and conn.DeleteSession() for deterministic
session management.
Channel numbers are now recycled immediately which prompted a fix for
TestSessionClose.
Fixed various tests to handle close frame (the error was being swallowed
before).
Tests that utilize testconn were silently failing due to a bug in
Conn.Read which has been fixed.

Fixes https://github.com/Azure/go-amqp/issues/164

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -33,10 +33,12 @@
...
 * Session will no longer flood peer with flow frames when half its incoming window is consumed.
+* Newly created `Session` won't leak if the context passed to `Client.NewSession()` expires before exit.
 
 ### Other Changes
 * Errors when reading/writing to the underlying `net.Conn` are now wrapped in a `ConnectionError` type.
 * Disambiguate error message for distinct cases where a session wasn't found for the specified remote channel.
 * Removed `link.Paused` as it didn't add much value and was broken in some cases.
 * Only send one flow frame when a drain has been requested.
 * Session window size increased to 5000.
+* Creation and deletion of `Session` instances have been made deterministic.
```
Comment (jhendrixMSFT) [inline CHANGELOG.md:44]: I'm going to look at moving this out of this PR so that it can go directly into `main`.

**Comment by richardpark-msft** [comment]
There's a failure in our websockets negotiation, I have a draft PR with my branch that merges in your changes:
https://github.com/Azure/azure-sdk-for-go/pull/18669

**Review by richardpark-msft — COMMENTED** [review]
Some intermediate comments, still reviewing.

File: conn.go [hunk 2]
```diff
@@ -299,316 +293,281 @@ func (c *conn) initTLSConfig() {
...
 
-		// create the next session to allocate
-		// note that channel always start at 0, and 0 is special and can't be deleted
-		nextChannel, _ = channels.Next()
-		nextSession    = newSessionResp{session: newSession(c, uint16(nextChannel))}
+	// create the next session to allocate
+	// note that channel always start at 0
+	channel, ok := c.channels.Next()
+	if !ok {
+		return nil, fmt.Errorf("reached connection channel max (%d)", c.channelMax)
```
Comment (richardpark-msft) [inline conn.go:401]: Nice to not ignore an error, unlikely as it might be!

File: conn.go [hunk 3]
```diff
@@ -299,316 +293,281 @@ func (c *conn) initTLSConfig() {
...
+	}
 }
 
 // Err returns the connection's error state after it's been closed.
 // Calling this on an open connection will block until the connection is closed.
 func (c *conn) Err() error {
-	c.errMu.Lock()
-	defer c.errMu.Unlock()
-	return &ConnectionError{inner: c.err}
+	<-c.Done
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply conn.go:389]: With this refactor, I might actually be able to safely remove `Err()`.  Looking...

File: conn.go [hunk 4]
```diff
@@ -299,316 +293,281 @@ func (c *conn) initTLSConfig() {
...
+			// this is the expected error when the connection is closed, swallow it
+			rxErr = nil
+		}
+	default:
+		// no pending read error
+	}
+
+	if txErr == nil && rxErr == nil && closeErr == nil {
+		// if there are no errors, it means user initiated close() and we shut down cleanly
+		c.doneErr = &ConnectionError{}
```
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply conn.go:373]: This is the error that's returned from APIs if the connection has been closed.  E.g. I close the connection then call `Sender.Send()` which should fail, returning this error.

File: conn.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhendrixMSFT) [reply conn.go]: Removed.

File: CHANGELOG.md (hunk 1 above)
Comment (jhendrixMSFT) [reply CHANGELOG.md:44]: Done, see https://github.com/Azure/go-amqp/pull/177

File: conn.go [hunk 5]
```diff
@@ -121,30 +121,29 @@ type conn struct {
...
 	channels            *bitmap.Bitmap
 	sessionsByChannel   map[uint16]*Session
 	sessionsByChannelMu sync.RWMutex
 
 	// connReader
-	rxProto       chan protoHeader  // protoHeaders received by connReader
-	rxFrame       chan frames.Frame // AMQP frames received by connReader
-	rxDone        chan struct{}
-	connReaderRun chan func() // functions to be run by conn reader (set deadline on conn to run)
+	rxBuf         buffer.Buffer // incomes bytes buffer
```
Comment (richardpark-msft) [inline conn.go:137]: `incoming`?

File: conn.go [hunk 6]
```diff
@@ -821,21 +747,56 @@ func (c *conn) exchangeProtoHeader(pID protoID) (stateFunc, error) {
...
-	case fr := <-c.rxFrame:
-		return p, fmt.Errorf("readProtoHeader: unexpected frame %#v", fr)
-	case <-deadline:
-		return p, errors.New("amqp: timeout waiting for response")
+
+	buf, ok := c.rxBuf.Next(protoHeaderSize)
+	if !ok {
+		return protoHeader{}, errors.New("invalid protoHeader")
 	}
+	_ = buf[7]
```
Comment (richardpark-msft) [inline conn.go:782]: Unneeded?

File: conn.go [hunk 7]
```diff
@@ -821,21 +747,56 @@ func (c *conn) exchangeProtoHeader(pID protoID) (stateFunc, error) {
...
 func (c *conn) readProtoHeader() (protoHeader, error) {
-	var deadline <-chan time.Time
-	if c.connectTimeout != 0 {
-		deadline = time.After(c.connectTimeout)
+	// only read from the network once our buffer has been exhausted.
+	// TODO: this preserves existing behavior as some tests rely on this
+	// implementation detail (it lets you replay a stream of bytes). we
+	// might want to consider removing this and fixing the tests as the
+	// protocol doesn't actually work this way.
+	if c.rxBuf.Len() == 0 {
```
Comment (richardpark-msft) [inline conn.go:755]: Is it possible to have _some_ bytes in rxBuf but just not enough to be a complete protocol header? If so, we should go in here if c.rxBuf.Len() < protoHeaderSize?

File: conn.go [hunk 8]
```diff
@@ -975,29 +936,25 @@ func (c *conn) saslOutcome() (stateFunc, error) {
...
-		return fr, errors.New("amqp: timeout waiting for response")
+	fr, err := c.readFrame()
+	if err != nil {
+		return frames.Frame{}, err
 	}
+
+	return fr, nil
 }
 
+const protoHeaderSize = 8
```
Comment (richardpark-msft) [inline conn.go:956]: NIT: Looks like you can just define this inside of readProtoHeader() and limit its scope a bit.

File: conn.go [hunk 9]
```diff
@@ -402,218 +387,160 @@ func (c *conn) DeleteSession(s *Session) {
...
-					break
-				}
+		if c.idleTimeout > 0 {
+			_ = c.net.SetReadDeadline(time.Now().Add(c.idleTimeout))
+		}
 
-				session.remoteChannel = fr.Channel
-				sessionsByRemoteChannel[fr.Channel] = session
+		var fr frames.Frame
+		fr, err = c.readFrame()
```
Comment (richardpark-msft) [inline conn.go:412]: NIT: f, err := c.readFrame()?

File: conn.go [hunk 10]
```diff
@@ -698,11 +621,14 @@ func (c *conn) connWriter() {
...
-			// send close
+		case <-c.rxtxExit:
+			// send close performative.  note that the spec says we
+			// SHOULD wait for the ack but we don't HAVE to, in order
+			// to be resilient to bad actors etc.  so we just send
+			// the close performative and exit.
 			cls := &frames.PerformClose{}
 			debug.Log(1, "TX (connWriter): %s", cls)
-			_ = c.writeFrame(frames.Frame{
+			c.txErr = c.writeFrame(frames.Frame{
```
Comment (richardpark-msft) [inline conn.go:631]: This is awesome, should help us see if we're getting desynched on closes!

**Review by richardpark-msft — COMMENTED** [review]
A few comments. There's enough here now that it's worth running the live tests against this version of go-amqp. I can walk you through that when you're ready.

File: conn.go (hunk 9 above)
Comment (jhendrixMSFT) [reply conn.go:412]: This is due to `err` being declared outside the loop and we don't want it to be shadowed.

File: conn.go (hunk 5 above)
Comment (jhendrixMSFT) [reply conn.go]: fixed

File: conn.go (hunk 6 above)
Comment (jhendrixMSFT) [reply conn.go]: fixed (unclear why this was originally there...)

File: conn.go (hunk 8 above)
Comment (jhendrixMSFT) [reply conn.go]: fixed

**Comment by jhendrixMSFT** [comment]
Test `TestLinkFlowWithDrain` hung.  Interestingly I can also repro this on `main`.

**Comment by jhendrixMSFT** [comment]
Test bug, fix pending.

File: conn.go (hunk 6 above)
Comment (jhendrixMSFT) [reply conn.go]: I think this is related to https://github.com/golang/go/issues/14808.  I see similar patterns in `encoding/binary` in the standard library.  I'm going to add this back with a comment.

**Comment by jhendrixMSFT** [comment]
Switching back to draft for now as I want to wait to merge this until after the API clean-up changes are in and released.

File: conn.go (hunk 7 above)
Comment (jhendrixMSFT) [reply conn.go:808]: We check `c.rxBuf.Len()` for the proper length in the `for` loop below and will keep reading as required.

File: conn.go [hunk 11]
```diff
@@ -454,251 +433,181 @@ func (c *Conn) deleteSession(s *Session) {
...
+			if !ok {
+				err = fmt.Errorf("%T: didn't find channel %d in sessionsByRemoteChannel", fr.Body, fr.Channel)
+				continue
+			}
+		}
 
-		// connection is complete
-		case <-c.closeMux:
+		select {
+		case session.rx <- fr:
```
Comment (richardpark-msft) [inline conn.go:516]: I'd be interesting to see how often we stall here. The 'session.rx' is an unbuffered channel so falling behind in one session could mean blocking the read loop altogther. Am I reading that incorrectly?

(also this is not a new problem, but it's just more clear how the code works since you stripped away more code!)

File: conn.go [hunk 12]
```diff
@@ -349,72 +340,60 @@ func (c *Conn) start() error {
...
+	// there was an error during shut-down or connReader/connWriter
+	// experienced a terminal error
+	return c.doneErr
 }
 
-// close should only be called by conn.mux.
+// close is called once, either from Close() or when connReader/connWriter exits
 func (c *Conn) close() {
-	close(c.done) // notify goroutines and blocked functions to exit
+	defer close(c.done)
```
Comment (richardpark-msft) [inline conn.go:366]: Could we move the `c.closeOnce.Do(func())` call into the `func (c *Conn) close()` function? Then you'd never have to worry about where or how often it's called.

File: conn.go [hunk 13]
```diff
@@ -349,72 +340,60 @@ func (c *Conn) start() error {
...
+
+	close(c.rxtxExit)
 
 	// wait for writing to stop, allows it to send the final close frame
 	<-c.txDone
```
Comment (richardpark-msft) [inline conn.go:371]: Is there a way to add a timeout to this? Or maybe we can set a write timeout for just the close frame when we transfer it.

File: conn.go [hunk 14]
```diff
@@ -349,72 +340,60 @@ func (c *Conn) start() error {
...
+		c.rxErr = nil
+	}
+
+	if c.txErr == nil && c.rxErr == nil && closeErr == nil {
+		// if there are no errors, it means user initiated close() and we shut down cleanly
+		c.doneErr = &ConnError{}
+	} else if amqpErr, ok := c.rxErr.(*Error); ok {
+		// we experienced a peer-initiated close that contained an Error.  return it
+		c.doneErr = &ConnError{RemoteErr: amqpErr}
+	} else if c.txErr != nil {
```
Comment (richardpark-msft) [inline conn.go:390]: I couldn't think of a reason but is there any significance to which err we check first? (txErr vs rxErr)

File: session.go [hunk 15]
```diff
@@ -42,9 +42,9 @@ type SessionOptions struct {
 // A session multiplexes Receivers.
 type Session struct {
 	channel       uint16                       // session's local channel
-	remoteChannel uint16                       // session's remote channel, owned by conn.mux
+	remoteChannel uint16                       // session's remote channel, owned by conn.connReader
 	conn          *Conn                        // underlying conn
-	rx            chan frames.Frame            // frames destined for this session are sent on this chan by conn.mux
+	rx            chan frames.Frame            // frames destined for this session are sent on this chan by conn.connReader
```
Comment (richardpark-msft) [inline session.go:47]: Is there any reason why this channel should not be buffered? I left a comment in the connReader() loop, but it seems like if a session falls behind it can clog up dispatching.

**Review by richardpark-msft — APPROVED** [review]
It's really looking good to me. It's in the heart of the AMQP stack though, so we should probably come up with some more extreme cases for testing as well since SB and EH don't necessarily push the boundaries too much in normal operation.

I also left some comments in spots where there might be perf bottlenecks that we could investigate as well.

File: conn.go (hunk 13 above)
Comment (jhendrixMSFT) [reply conn.go:371]: I have such a change (and more!) pending in my [reliable_writes](https://github.com/jhendrixMSFT/go-amqp/tree/reliable_writes) branch (it builds on this PR).

File: conn.go (hunk 11 above)
Comment (jhendrixMSFT) [reply conn.go:516]: Your reading is correct and is something I've been pondering on for some time now.  Unfortunately, removing the session mux is much more complicated than conn's mux as all of the flow control is managed there and depends on proper synchronization of incoming and outgoing frames (I've taken a few stabs at it, it's not pretty).

File: conn.go (hunk 14 above)
Comment (jhendrixMSFT) [reply conn.go:390]: Yes, but only insofar as what error message is exposed to the caller via the `ConnError` in the event that `connReader()` and `connWriter()` experienced terminal errors.

I debated about concatenating them both together but decided against it as I felt the added complexity (which admittingly is minor) wasn't worth it (i.e. how actionable would the error message really be).  I could be convinced to do this though if you think there's a compelling reason.

File: session.go (hunk 15 above)
Comment (jhendrixMSFT) [reply session.go:47]: I believe this needs to be split into two channels, one for transfer frames and one for all others.  The transfer frames channel should be buffered to the size of our incoming window.

For the non-transfer frames channel, I don't _think_ it needs to be buffered but I haven't had a chance to dig into it deeply enough.  The part that can cause clogs (or worse), and that I want to fix in general, is when the link is writing to its session mux and has to "pump" it.

File: conn.go (hunk 12 above)
Comment (jhendrixMSFT) [reply conn.go]: Good idea, fixed.

File: conn.go (hunk 6 above)
Comment (richardpark-msft) [reply conn.go:827]: Would it work equally well to do _ = buf[protoHeaderSize-1], just to show that's where the '7' is coming from?

Comment (jhendrixMSFT) [reply conn.go]: done

**Review by richardpark-msft — APPROVED** [review]
